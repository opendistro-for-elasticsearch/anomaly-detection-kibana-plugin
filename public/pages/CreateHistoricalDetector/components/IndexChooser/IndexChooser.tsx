/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { EuiCallOut, EuiComboBox, EuiSpacer } from '@elastic/eui';
import { Field, FieldProps, FormikProps } from 'formik';
import { debounce, get } from 'lodash';
import React, { useEffect, useState, Fragment } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { CatIndex, IndexAlias } from '../../../../../server/models/types';
import ContentPanel from '../../../../components/ContentPanel/ContentPanel';
import { AppState } from '../../../../redux/reducers';
import {
  getIndices,
  getMappings,
  getPrioritizedIndices,
} from '../../../../redux/reducers/elasticsearch';
import { FormattedFormRow } from '../../../createDetector/components/FormattedFormRow/FormattedFormRow';
import { getVisibleOptions, sanitizeSearchText } from '../../../utils/helpers';
import { getError, isInvalid } from '../../../../utils/utils';
import { IndexOption } from '../../../createDetector/components/Datasource/IndexOption';
import { validateIndex } from '../../../utils/validate';
import {
  HistoricalDetectorFormikValues,
  INITIAL_HISTORICAL_DETECTOR_VALUES,
} from '../../utils/constants';

interface IndexChooserProps {
  formikProps: FormikProps<HistoricalDetectorFormikValues>;
  isEdit: boolean;
}

export function IndexChooser(props: IndexChooserProps) {
  const dispatch = useDispatch();
  const elasticsearchState = useSelector(
    (state: AppState) => state.elasticsearch
  );
  const [queryText, setQueryText] = useState('');

  useEffect(() => {
    const getInitialIndices = async () => {
      await dispatch(getIndices(queryText));
    };
    getInitialIndices();
  }, []);

  const handleSearchChange = debounce(async (searchValue: string) => {
    if (searchValue !== queryText) {
      const sanitizedQuery = sanitizeSearchText(searchValue);
      setQueryText(sanitizedQuery);
      await dispatch(getPrioritizedIndices(sanitizedQuery));
    }
  }, 300);

  const handleIndexNameChange = (selectedOptions: any) => {
    const indexName = get(selectedOptions, '0.label', '');
    if (indexName !== '') {
      dispatch(getMappings(indexName));
    }
  };

  const visibleIndices = get(elasticsearchState, 'indices', []) as CatIndex[];
  const visibleAliases = get(elasticsearchState, 'aliases', []) as IndexAlias[];

  return (
    <ContentPanel title="Data source" titleSize="s">
      {props.isEdit ? (
        <div>
          <EuiCallOut
            title="Modifying the index selected will reset your detector configuration settings."
            color="warning"
            iconType="alert"
            size="s"
          />
          <EuiSpacer />
        </div>
      ) : null}
      <Fragment>
        <Field name="index" validate={validateIndex}>
          {({ field, form }: FieldProps) => {
            return (
              <FormattedFormRow
                title="Index"
                hint="Choose an index or index pattern as the data source."
                isInvalid={isInvalid(field.name, form)}
                error={getError(field.name, form)}
                helpText="You can use a wildcard (*) in your index pattern"
              >
                <EuiComboBox
                  id="index"
                  placeholder="Find indices"
                  async
                  isLoading={elasticsearchState.requesting}
                  options={getVisibleOptions(visibleIndices, visibleAliases)}
                  onSearchChange={handleSearchChange}
                  onCreateOption={(createdOption: string) => {
                    const normalizedOptions = createdOption.trim();
                    if (!normalizedOptions) return;
                    const customOption = [{ label: normalizedOptions }];
                    form.setFieldValue('index', customOption);
                    handleIndexNameChange(customOption);
                  }}
                  onBlur={() => {
                    form.setFieldTouched('index', true);
                  }}
                  onChange={(options) => {
                    const newOption = get(options, '0.label');
                    const prevOption = get(field, 'value.0.label');
                    if (newOption !== prevOption) {
                      form.setFieldValue('index', options);
                      form.setFieldValue('timeField', undefined);
                      form.setFieldValue(
                        'featureList',
                        INITIAL_HISTORICAL_DETECTOR_VALUES.featureList
                      );
                      form.setFieldValue(
                        'detectionInterval',
                        INITIAL_HISTORICAL_DETECTOR_VALUES.detectionInterval
                      );
                      handleIndexNameChange(options);
                    }
                  }}
                  selectedOptions={field.value}
                  singleSelection={true}
                  isClearable={true}
                  renderOption={(option, searchValue, className) => (
                    <IndexOption
                      option={option}
                      searchValue={searchValue}
                      contentClassName={className}
                    />
                  )}
                />
              </FormattedFormRow>
            );
          }}
        </Field>
      </Fragment>
    </ContentPanel>
  );
}
