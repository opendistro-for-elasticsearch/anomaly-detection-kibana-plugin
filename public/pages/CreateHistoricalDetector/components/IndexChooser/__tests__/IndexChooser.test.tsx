/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { Provider } from 'react-redux';
import {
  HashRouter as Router,
  RouteComponentProps,
  Route,
  Switch,
} from 'react-router-dom';
import { render, fireEvent, wait } from '@testing-library/react';
import { IndexChooser } from '../IndexChooser';
import configureStore from '../../../../../redux/configureStore';
import { httpClientMock, coreServicesMock } from '../../../../../../test/mocks';
import { CoreServicesContext } from '../../../../../components/CoreServices/CoreServices';
import { Formik } from 'formik';

const CALLOUT_TEXT =
  'Modifying the selected index resets your detector configuration.';
const HELP_TEXT = 'Choose an index or index pattern as the data source.';
const INVALID_TEXT = 'Must specify an index';
const PLACEHOLDER_TEXT = 'Find indices';

const renderWithRouter = (isEdit: boolean = false) => ({
  ...render(
    <Provider store={configureStore(httpClientMock)}>
      <Router>
        <Switch>
          <Route
            render={(props: RouteComponentProps) => (
              <CoreServicesContext.Provider value={coreServicesMock}>
                <Formik initialValues={{}} onSubmit={jest.fn()}>
                  {() => (
                    <div>
                      <IndexChooser
                        formikProps={{ values: {} }}
                        isEdit={isEdit}
                        {...props}
                      />
                    </div>
                  )}
                </Formik>
              </CoreServicesContext.Provider>
            )}
          />
        </Switch>
      </Router>
    </Provider>
  ),
});

describe('<IndexChooser /> spec', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  test('renders the component in create mode', () => {
    httpClientMock.get = jest.fn().mockResolvedValue({
      ok: true,
      response: {
        indices: [
          { index: 'hello', health: 'green' },
          { index: 'world', health: 'yellow' },
        ],
      },
    });
    const { container, getByText, queryByText } = renderWithRouter(false);
    expect(container.firstChild).toMatchSnapshot();
    getByText(HELP_TEXT);
    expect(queryByText(CALLOUT_TEXT)).toBeNull();
  });

  test('renders the component in edit mode with callout', () => {
    httpClientMock.get = jest.fn().mockResolvedValue({
      ok: true,
      response: {
        indices: [
          { index: 'hello', health: 'green' },
          { index: 'world', health: 'yellow' },
        ],
      },
    });
    const { container, getByText, findByText } = renderWithRouter(true);
    expect(container.firstChild).toMatchSnapshot();
    getByText(HELP_TEXT);
    expect(findByText(CALLOUT_TEXT)).not.toBeNull();
  });
  test('shows error for invalid index selection', async () => {
    console.warn = jest.fn();
    const { queryByText, findByText, getByText } = renderWithRouter(false);
    expect(queryByText(INVALID_TEXT)).toBeNull();
    fireEvent.focus(getByText(PLACEHOLDER_TEXT));
    await wait();
    fireEvent.blur(getByText(PLACEHOLDER_TEXT));
    await wait();
    expect(findByText(INVALID_TEXT)).not.toBeNull();
  });
});
