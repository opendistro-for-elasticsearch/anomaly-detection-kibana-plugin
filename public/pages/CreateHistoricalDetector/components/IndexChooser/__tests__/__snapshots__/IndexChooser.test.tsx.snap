// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IndexChooser /> spec renders the component in create mode 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Data source
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 10px 0px;"
    >
      <div
        class="euiFormRow"
        hint="Choose an index or index pattern as the data source."
        id="random_html_id-row"
        title="Index"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            aria-invalid="false"
            class="euiFormLabel euiFormRow__label"
            for="random_html_id"
          >
            <div
              style="line-height: 8px;"
            >
              <p>
                Index
              </p>
              <br />
              <p
                class="sublabel"
              >
                Choose an index or index pattern as the data source.
              </p>
            </div>
          </label>
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            aria-describedby="random_html_id-help"
            aria-expanded="false"
            aria-haspopup="listbox"
            class="euiComboBox"
            role="combobox"
          >
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <div
                  class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isLoading euiComboBox__inputWrap-isClearable"
                  data-test-subj="comboBoxInput"
                  tabindex="-1"
                >
                  <p
                    class="euiComboBoxPlaceholder"
                  >
                    Find indices
                  </p>
                  <div
                    class="euiComboBox__input"
                    style="font-size: 14px; display: inline-block;"
                  >
                    <input
                      aria-controls=""
                      data-test-subj="comboBoxSearchInput"
                      id="random_html_id"
                      role="textbox"
                      style="box-sizing: content-box; width: 2px;"
                      value=""
                    />
                    <div
                      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                    />
                  </div>
                </div>
                <div
                  class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                >
                  <span
                    class="euiLoadingSpinner euiLoadingSpinner--medium"
                  />
                  <button
                    aria-label="Open list of options"
                    class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                    data-test-subj="comboBoxToggleListButton"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFormHelpText euiFormRow__text"
            id="random_html_id-help"
          >
            You can use a wildcard (*) in your index pattern
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<IndexChooser /> spec renders the component in edit mode with callout 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Data source
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 10px 0px;"
    >
      <div>
        <div
          class="euiCallOut euiCallOut--warning euiCallOut--small"
        >
          <div
            class="euiCallOutHeader"
          >
            <svg
              aria-hidden="true"
              class="euiIcon euiIcon--medium euiIcon-isLoading euiCallOutHeader__icon"
              focusable="false"
              height="16"
              role="img"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            />
            <span
              class="euiCallOutHeader__title"
            >
              Modifying the index selected will reset your detector configuration settings.
            </span>
          </div>
        </div>
        <div
          class="euiSpacer euiSpacer--l"
        />
      </div>
      <div
        class="euiFormRow"
        hint="Choose an index or index pattern as the data source."
        id="random_html_id-row"
        title="Index"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            aria-invalid="false"
            class="euiFormLabel euiFormRow__label"
            for="random_html_id"
          >
            <div
              style="line-height: 8px;"
            >
              <p>
                Index
              </p>
              <br />
              <p
                class="sublabel"
              >
                Choose an index or index pattern as the data source.
              </p>
            </div>
          </label>
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            aria-describedby="random_html_id-help"
            aria-expanded="false"
            aria-haspopup="listbox"
            class="euiComboBox"
            role="combobox"
          >
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <div
                  class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isLoading euiComboBox__inputWrap-isClearable"
                  data-test-subj="comboBoxInput"
                  tabindex="-1"
                >
                  <p
                    class="euiComboBoxPlaceholder"
                  >
                    Find indices
                  </p>
                  <div
                    class="euiComboBox__input"
                    style="font-size: 14px; display: inline-block;"
                  >
                    <input
                      aria-controls=""
                      data-test-subj="comboBoxSearchInput"
                      id="random_html_id"
                      role="textbox"
                      style="box-sizing: content-box; width: 2px;"
                      value=""
                    />
                    <div
                      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                    />
                  </div>
                </div>
                <div
                  class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                >
                  <span
                    class="euiLoadingSpinner euiLoadingSpinner--medium"
                  />
                  <button
                    aria-label="Open list of options"
                    class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                    data-test-subj="comboBoxToggleListButton"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M13.069 5.157L8.384 9.768a.546.546 0 01-.768 0L2.93 5.158a.552.552 0 00-.771 0 .53.53 0 000 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 000-.76.552.552 0 00-.771 0z"
                        fill-rule="non-zero"
                      />
                    </svg>
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFormHelpText euiFormRow__text"
            id="random_html_id-help"
          >
            You can use a wildcard (*) in your index pattern
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
